LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaApiHttpClient.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Net;
+    using System.Net.Http;
+    using LoRaTools.CommonAPI;
+    using LoRaWan.Core;
+    using Microsoft.Extensions.DependencyInjection;
+    using Polly;
+
+    public static class LoRaApiHttpClient
+    {
+        public const string Name = nameof(LoRaApiHttpClient);
+    }
+
+    public static class LoRaApiHttpClientExtensions
+    {
+        private const int NumberOfAggressiveRetries = 4;
+        private const int NumberOfRetries = 8;
+        private static readonly TimeSpan AggressiveRetryInterval = TimeSpan.FromMilliseconds(50);
+        private static readonly TimeSpan RetryInterval = TimeSpan.FromMilliseconds(120);
+
+        public static IServiceCollection AddApiClient(this IServiceCollection services,
+                                                      NetworkServerConfiguration configuration,
+                                                      ApiVersion expectedFunctionVersion) =>
+            AddApiClient(services, () =>
+            {
+                var handler = new ServiceFacadeHttpClientHandler(expectedFunctionVersion);
+
+                if (!string.IsNullOrEmpty(configuration.HttpsProxy))
+                {
+                    var webProxy = new WebProxy(
+                        new Uri(configuration.HttpsProxy),
+                        BypassOnLocal: false);
+
+                    handler.Proxy = webProxy;
+                    handler.UseProxy = true;
+                }
+
+                return handler;
+            });
+
+        internal static IServiceCollection AddApiClient(this IServiceCollection services, Func<HttpMessageHandler> createHttpMessageHandler)
+        {
+            _ = services.AddHttpClient(LoRaApiHttpClient.Name)
+                        .ConfigurePrimaryHttpMessageHandler(createHttpMessageHandler)
+                        // Retry aggressively first to not miss the receive window if possible.
+                        .AddTransientHttpErrorPolicy(policyBuilder =>
+                            policyBuilder.WaitAndRetryAsync(NumberOfRetries, i => (i <= NumberOfAggressiveRetries ? AggressiveRetryInterval : RetryInterval) * Math.Pow(2, i)));
```

Comment:
For the reviewer: Since all messages of the same device are processed in series, and the retries as defined here might take up to a minute, it's possible that the request queue is blocked for too long. Should we make this shorter?

Reply:
Not a very educated opinion here but 8 retries in total sound a lot, maybe 6 (3 of them aggressive) and shorter intervals would be good enough?

Reply:
I think that's ok. We will miss the downstream window potentially, but for any unconfirmed messages up, we get higher deliver guarantee.